src/compiler/checker.ts

```diff
@@ -13174,8 +13174,9 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         // and T as the template type.
         const typeParameter = getTypeParameterFromMappedType(type);
         const constraintType = getConstraintTypeFromMappedType(type);
-        const nameType = getNameTypeFromMappedType(type.target as MappedType || type);
-        const isFilteringMappedType = nameType && isTypeAssignableTo(nameType, typeParameter);

```

Comment:
the bug was here, this likely should have been something like this:
```diff
-const isFilteringMappedType = nameType && isTypeAssignableTo(nameType, typeParameter);
-const isFilteringMappedType = nameType && isTypeAssignableTo(nameType, getTypeParameterFromMappedType(type.target as MappedType || type));
```

and this is what happens in the added `isFilteringMappedType` since it unpacks the correct mpped type on its own and uses it as the source of information for both arguments passed to `isTypeAssignableTo`